src/finiteVolume/processes/soot.cpp

```diff
@@ -0,0 +1,397 @@
+    PetscCall(VecGetArray(fVec, &fArray));
+
+    // get access to the source array in the solver
+    const PetscScalar* sourceArray;
+    PetscCall(VecGetArrayRead(soot->sourceVec, &sourceArray));
+
+    // get the dm
+    // March over each cell
+    for (PetscInt c = cellRange.start; c < cellRange.end; ++c) {
+        // if there is a cell array, use it, otherwise it is just c
+        const PetscInt cell = cellRange.GetPoint(c);
+
+        // read the global f
+        PetscScalar* rhs;
+        PetscCall(DMPlexPointLocalRef(dm, cell, fArray, &rhs));
+
+        // if a real cell
+        if (rhs) {
+            // read the source from the local calc
+            const PetscScalar* source;
+            PetscCall(DMPlexPointLocalRead(soot->sourceDm, cell, sourceArray, &source));
+
+            // add the yi sources
+            for (PetscInt s = 0; s < TOTAL_ODE_SPECIES; ++s) {
+                rhs[soot->pointInformation.speciesOffset[s]] += source[s];
+            }
+            // Add in the energy source, it was stored in the temperature location
+            rhs[flowEulerId.offset + ablate::finiteVolume::CompressibleFlowFields::RHOE] += source[ODE_T];
+            rhs[flowDensityProgressId.offset] += source[ODE_NDD];
+        }
+    }
+
+    solver.RestoreRange(cellRange);
+    PetscCall(VecRestoreArray(fVec, &fArray));
+    PetscCall(VecGetArrayRead(soot->sourceVec, &sourceArray));
+
+    PetscFunctionReturn(0);
+}
+PetscErrorCode ablate::finiteVolume::processes::Soot::SinglePointSootChemistryRHS(TS ts, PetscReal t, Vec xVec, Vec fVec, void* ctx) {
+    PetscFunctionBeginUser;
+    auto pointInfo = (OdePointInformation*)ctx;
+
+    // extract the read/write arrays
+    PetscScalar* xArray;
+    PetscCall(VecGetArray(xVec, &xArray));
+    PetscCall(VecZeroEntries(fVec));
+    PetscScalar* fArray;
+    PetscCall(VecGetArray(fVec, &fArray));
+
+    // copy over the updated to the scratch variable for now
+    for (std::size_t s = 0; s < TOTAL_ODE_SPECIES; s++) {
+        xArray[s] = PetscMax(PetscMin(xArray[s], 1.0), 0.0);
+        pointInfo->yiScratch[pointInfo->speciesIndex[s]] = xArray[s];
+    }
+    xArray[ODE_T] = PetscMax(xArray[ODE_T], 0);
+    xArray[ODE_NDD] = PetscMax(xArray[ODE_NDD], 0);
+
+    // Add in the Soot Reaction Sources
+    real_type SVF = xArray[C_s] * pointInfo->currentDensity / solidCarbonDensity;
+
+    // Total S.A. of soot / unit volume
+    PetscReal SA_V = calculateSurfaceArea_V(xArray[C_s], xArray[ODE_NDD], pointInfo->currentDensity);
+
+    // Need the Concentrations of C2H2, O2, O, and OH
+    // It is unclear in the formulations of the Reaction Rates whether to use to concentration in regards to the total mixture or just the gas phace, There is a difference due to the density relation
+    //-> For now we will use the concentration to be the concentration in the gas phase as it makes more physical sense
+    PetscReal C2H2Conc = pointInfo->currentDensity * PetscMax(0, xArray[C2H2]) / pointInfo->mw[C2H2];
```

Comment:
Have you considered making these functions and the stuff after them children of an interface that can be iterated over? Just so it might look cleaner.

Reply:
Yeah I think that is a good idea.  This was a direct port of existing code, if we are happy with the results we should go back and clean this up.